src/serializer/ResidualHeapSerializer.js

```diff
@@ -619,40 +614,77 @@ export class ResidualHeapSerializer {
         ).length;
       }
 
-      if (numAdditionalFunctionReferences > 0 || !this._options.delayInitializations || this._options.simpleClosures) {
-        // We can just emit it into the current function body.
-        return {
-          body: this.currentFunctionBody,
-          usedOnlyByAdditionalFunctions: numAdditionalFunctionReferences === functionValues.length,
-          description: "this.currentFunctionBody",
-        };
-      } else {
+      if (
+        numAdditionalFunctionReferences === 0 &&
+        this._options.delayInitializations &&
+        !this._options.simpleClosures
+      ) {
         // We can delay the initialization, and move it into a conditional code block in the residual functions!
         let body = this.residualFunctions.residualFunctionInitializers.registerValueOnlyReferencedByResidualFunctions(
           functionValues,
           val
         );
         return { body, usedOnlyByResidualFunctions: true, description: "delay_initializer" };
       }
+
+      usedOnlyByAdditionalFunctions = numAdditionalFunctionReferences === functionValues.length;
     }
 
+    let getBody = s => {
+      if (s === this.generator || s instanceof FunctionValue) {
+        return this.currentFunctionBody;
+      } else {
+        return this.activeGeneratorBodies.get(s);
+      }
+    };
+
     // This value is referenced from more than one generator or function.
-    // We can emit the initialization of this value into the body associated with their common ancestor.
-    let commonAncestor = Array.from(scopes).reduce((x, y) => commonAncestorOf(x, y), generators[0]);
-    invariant(commonAncestor instanceof Generator); // every scope is either the root, or a descendant
+    // Let's find the body associated with their common ancestor.
+    let commonAncestor = Array.from(scopes).reduce((x, y) => commonAncestorOf(x, y), scopes.values().next().value);
```

Comment:
It is not obvious to me what scopes.values().next().value is at this point, so a comment seems called for.

Reply:
It's the "first scope" element. Will clarify.